evdns.c

```diff
@@ -1480,9 +1492,49 @@ request_parse(u8 *packet, int length, struct evdns_server_port *port,
 		server_req->base.questions[server_req->base.nquestions++] = q;
 	}
 
-	/* Ignore answers, authority, and additional. */
+#define SKIP_RR \
+	do { \
+		SKIP_NAME; \
+		j += 2 /* type */ + 2 /* class */ + 4 /* ttl */; \
+		GET16(rdlen); \
+		j += rdlen; \
+	} while (0)
+
+	for (i = 0; i < answers; ++i) {
+		SKIP_RR;
+	}
+
+	for (i = 0; i < authority; ++i) {
+		SKIP_RR;
+	}
+
+	server_req->max_udp_reply_size = DNS_MAX_UDP_SIZE;
+	for (i = 0; i < additional; ++i) {
+		SKIP_NAME;
+		GET16(type);
+		GET16(class);
+		GET32(ttl);
+		GET16(rdlen);
+		(void)ttl;
+		j += rdlen;
+		if (type == TYPE_OPT) {
+			/* In case of OPT pseudo-RR `class` field is treated
+			 * as a requestor's UDP payload size. */
+			server_req->max_udp_reply_size = MAX(class, DNS_MAX_UDP_SIZE);
+			evdns_server_request_add_reply(&(server_req->base),
+				EVDNS_ADDITIONAL_SECTION,
+				"", /* name */
+				TYPE_OPT, /* type */
+				DNS_MAX_UDP_SIZE, /* class */
+				0, /* ttl */
+				0, /* datalen */
+				0, /* is_name */
+				NULL /* data */
+			);
```

Comment:
what is the purpose of `evdns_server_request_add_reply`?

Reply:
It's part of https://tools.ietf.org/html/rfc6891#section-6.1.1
```
If an OPT record is present in a received request, compliant
responders MUST include an OPT record in their respective responses.
```
Here is an example request to Google Public DNS:
```
dig @8.8.8.8 +edns=0 +bufsize=12024 msomw23.ru
```
Here is an answer:
![image](https://user-images.githubusercontent.com/16498414/83461506-99fd5e80-a471-11ea-9a27-2274c0a36274.png)
As you see it contains this special additional section.

Reply:
Maybe it's worth adding a comment in the code?

Reply:
I just missed the place where it added into the response, that's why I asked, found it now